File: ProjectTest/OneLang/CSharp/One/Transforms/ResolveUnresolvedTypes.cs
using One;
using One.Ast;

namespace One.Transforms
{
    public class ResolveUnresolvedTypes : AstTransformer {
        public ResolveUnresolvedTypes(): base("ResolveUnresolvedTypes")
        {
            
        }
        
        protected override Type_ visitType(Type_ type) {
            base.visitType(type);
            if (type is UnresolvedType unrType) {
                var symbol = this.currentFile.availableSymbols.get(unrType.typeName);
                if (symbol == null) {
                    this.errorMan.throw_($"Unresolved type '{unrType.typeName}' was not found in available symbols");
                    return null;
                }
                
                if (symbol is Class class_)
                    return new ClassType(class_, unrType.typeArguments);
                else if (symbol is Interface int_)
                    return new InterfaceType(int_, unrType.typeArguments);
                else if (symbol is Enum_ enum_)
                    return new EnumType(enum_);
                else {
                    this.errorMan.throw_($"Unknown symbol type: {symbol}");
                    return null;
                }
            }
            else
                return null;
        }
        
        protected override Expression visitExpression(Expression expr) {
            if (expr is UnresolvedNewExpression unrNewExpr) {
                var clsType = this.visitType(unrNewExpr.cls);
                if (clsType is ClassType classType) {
                    var newExpr = new NewExpression(classType, unrNewExpr.args);
                    newExpr.parentNode = unrNewExpr.parentNode;
                    base.visitExpression(newExpr);
                    return newExpr;
                }
                else {
                    this.errorMan.throw_($"Excepted ClassType, but got {clsType}");
                    return null;
                }
            }
            else
                return base.visitExpression(expr);
        }
    }
}

File: ProjectTest/OneLang/Java/src/main/java/OneLang/One/Transforms/ResolveUnresolvedTypes/ResolveUnresolvedTypes.java
import java.util.Arrays;

public class ResolveUnresolvedTypes extends AstTransformer {
    public ResolveUnresolvedTypes()
    {
        super("ResolveUnresolvedTypes");
        
    }
    
    protected IType visitType(IType type) {
        super.visitType(type);
        if (type instanceof UnresolvedType) {
            if (this.currentInterface != null && Arrays.stream(this.currentInterface.getTypeArguments()).anyMatch(((UnresolvedType)type).typeName::equals))
                return new GenericsType(((UnresolvedType)type).typeName);
            
            var symbol = this.currentFile.availableSymbols.get(((UnresolvedType)type).typeName);
            if (symbol == null) {
                this.errorMan.throw_("Unresolved type '" + ((UnresolvedType)type).typeName + "' was not found in available symbols");
                return null;
            }
            
            if (symbol instanceof Class)
                return new ClassType(((Class)symbol), ((UnresolvedType)type).getTypeArguments());
            else if (symbol instanceof Interface)
                return new InterfaceType(((Interface)symbol), ((UnresolvedType)type).getTypeArguments());
            else if (symbol instanceof Enum)
                return new EnumType(((Enum)symbol));
            else {
                this.errorMan.throw_("Unknown symbol type: " + symbol);
                return null;
            }
        }
        else
            return null;
    }
    
    protected Expression visitExpression(Expression expr) {
        if (expr instanceof UnresolvedNewExpression) {
            var clsType = this.visitType(((UnresolvedNewExpression)expr).cls);
            if (clsType instanceof ClassType) {
                var newExpr = new NewExpression(((ClassType)clsType), ((UnresolvedNewExpression)expr).args);
                newExpr.parentNode = ((UnresolvedNewExpression)expr).parentNode;
                super.visitExpression(newExpr);
                return newExpr;
            }
            else {
                this.errorMan.throw_("Excepted ClassType, but got " + clsType);
                return null;
            }
        }
        else
            return super.visitExpression(expr);
    }
}